site/app/models/User.php

```diff
@@ -233,4 +233,39 @@ public function getAnonId() {
         }
         return $this->anon_id;
     }
+
+    /**
+     * Checks $data to make sure it is acceptable for $field.
+     *
+     * @param string $field
+     * @param mixed $data
+     * @return bool
+     */
+    static public function validateUserData($field, $data) {
+
+    	switch($field) {
+		case 'user_id':
+			//Username / useer_id must contain only lowercase alpha, numbers, underscores, hyphens
+			return preg_match("~^[a-z0-9_\-]+$~", $data) === 1;
+		case 'user_firstname':
+		case 'user_lastname':
+		case 'user_preferred_firstname':
+			//First, Last, Preferred name must be alpha characters, white-space, or certain punctuation.
+        	return preg_match("~^[a-zA-Z'`\-\. ]+$~", $data) === 1;
+		case 'user_email':
+			//Check email address for appropriate format. e.g. "user@university.edu", "user@cs.university.edu", etc.
+			return preg_match("~^[^(),:;<>@\\\"\[\]]+@(?!\-)[a-zA-Z0-9\-]+(?<!\-)(\.[a-zA-Z0-9]+)+$~", $data) === 1;
+		case 'user_group':
+            //user_group check is a digit between 1 - 4.
+			return preg_match("~^[1-4]{1}$~", $data) === 1;
+		case 'user_password':
+	        //Database password cannot be blank, no check on format
+			return $data !== "";
+		default:
+			//$data can't be validated since $field is unknown.  Notify developer with a stop error (also protectes data record integrity).
+			$field = var_export(htmlentities($field), true);
+			$data = var_export(htmlentities($data), true);
+			trigger_error('User::validateUserData() called with unknown $field '.$field.' and $data '.$data, E_USER_ERROR);
```

Comment:
Let's throw an exception here as that gets logged in a better fashion than fatal triggered errors (which are logged, but also weirdly displayed to users).

Reply:
I tried `throw new Exception()`, but that always popped up as `apps/models/Exception is not defined`.  Pretty much the same when I tried `throw new Error()`.  I didn't think that was helpful, so I went with `trigger_error()` to be more descriptive.